### PR TITLE
Explicitly set Bitbucket Server provider name, fix GitHub url parser

### PR DESCRIPTION
**Pull request by vinokurig**
### What does this PR do?
* Return 'bitbucket-server' on oauth list request if the provider is Bitbucket server. If the provider is Bitbucket SAAS, return `bitbucket`.
* Do not test GitHub repository urls with user request if the url is a GitHub SAAS url.

### What issues does this PR fix or reference?

https://github.com/eclipse-che/che/issues/22971

### How to test this PR?

1. Deploy che with the PR image: `quay.io/eclipse/che-server:pr-689`
2. Configure BitBucket server oauth2.
3. Call `<che-endpoint/api/oauth> api request, see: the provider name is `bitbucket-server`
4. Re-configure the oauth secret to BitBucket SAAS.
5. Call `<che-endpoint/api/oauth> api request, see: the provider name is `bitbucket`

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [ ] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [ ] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [ ] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [ ] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [ ] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: wsmaster/che-core-api-auth-bitbucket/src/main/java/org/eclipse/che/security/oauth/BitbucketOAuthAuthenticator.java [hunk 1]
```diff
@@ -52,7 +54,7 @@ public String getAuthenticateUrl(URL requestUrl, List<String> scopes) {
 
   @Override
   public final String getOAuthProvider() {
-    return "bitbucket";
+    return BITBUCKET_CLOUD_ENDPOINT.equals(bitbucketEndpoint) ? "bitbucket" : "bitbucket-server";
```
Comment (ibuziuk) [inline wsmaster/che-core-api-auth-bitbucket/src/main/java/org/eclipse/che/security/oauth/BitbucketOAuthAuthenticator.java:57]: should we extract consts as well for bitbucket / bitbucket-server ? would be nice to add comments java docs as well

**Review by ibuziuk — APPROVED** [review]
minor comments about consts extractions

File: wsmaster/che-core-api-auth-bitbucket/src/main/java/org/eclipse/che/security/oauth/BitbucketOAuthAuthenticator.java (hunk 1 above)
Comment (vinokurig) [reply wsmaster/che-core-api-auth-bitbucket/src/main/java/org/eclipse/che/security/oauth/BitbucketOAuthAuthenticator.java]: done

**Review by tolusha — APPROVED** [review]

File: wsmaster/che-core-api-factory-github-common/src/main/java/org/eclipse/che/api/factory/server/github/AbstractGithubURLParser.java [hunk 2]
```diff
@@ -93,8 +93,8 @@ public boolean isValid(@NotNull String url) {
         // If the GitHub URL is not configured, try to find it in a manually added user namespace
         // token.
         || isUserTokenPresent(trimmedUrl)
-        // Try to call an API request to see if the URL matches GitHub.
-        || isApiRequestRelevant(trimmedUrl);
+        // Try to call an API request to see if the URL matches self-hosted GitHub Enterprise.
+        || (!GITHUB_SAAS_ENDPOINT.equals(endpoint) && isApiRequestRelevant(trimmedUrl));
```
[the remark this answers is not in the document]
Comment (vinokurig) [reply wsmaster/che-core-api-factory-github-common/src/main/java/org/eclipse/che/api/factory/server/github/AbstractGithubURLParser.java:103]: > Can we rename isApiRequestRelevant to isGitHubServerURLValid ?

I think that `isApiRequestRelevant` is a more specific name. We already validate the GitHub Server url on each step: 
1. The GitHub Server url might be configured so the regex will detect it.
2. A user token secret from the GitHub Server might be present.
3. If all the checks above returned `false`, we finally send a test request.

[the remark this answers is not in the document]
Comment (vinokurig) [reply wsmaster/che-core-api-factory-github-common/src/main/java/org/eclipse/che/api/factory/server/github/AbstractGithubURLParser.java:103]: Added the javadocs to the corresponding steps.

**Comment by artaleks9** [comment]
Verified on Eclipse Che with `quay.io/eclipse/che-server:pr-689` - **the functionality works properly**

Responses on the `<che-endpoint/api/oauth>` API requests:
```
[{"name":"bitbucket-server","endpointUrl":"https://bitbucket-bitbucket.apps.ds-airgap-v16.crw-qe.com","links":[{"method":"GET","parameters":[{"name":"oauth_provider","defaultValue":"bitbucket-server","required":true,"valid":[]},{"name":"mode","defaultValue":"federated_login","required":true,"valid":[]}],"rel":"Authenticate URL","href":"https://eclipse-che.apps.cluster-hx4vv.dynamic.redhatworkshops.io/api/oauth/authenticate"}]}]

-------

[{"name":"bitbucket","endpointUrl":"https://bitbucket.org","links":[{"method":"GET","parameters":[{"name":"oauth_provider","defaultValue":"bitbucket","required":true,"valid":[]},{"name":"mode","defaultValue":"federated_login","required":true,"valid":[]}],"rel":"Authenticate URL","href":"https://eclipse-che.apps.cluster-hx4vv.dynamic.redhatworkshops.io/api/oauth/authenticate"}]}]
```

**Review by artaleks9 — APPROVED** [review]

**Pull request merged**